### ISSUE TITLE
HyperlinkCell cannot be selected with mouse in order to edit

### ISSUE THREAD

**Issue by lamazze** [unvell/ReoGrid#309]
- Set the cell type to HyperlinkCell
`sheet.ColumnHeaders[0].DefaultCellBody = typeof(HyperlinkCell);
- Load data in the first column
- Select with the mouse Cell[1,1]
- Try to select Cell[0,0] with mouse in order to edit it. You can't. Obviously the

**Comment by lamazze** [comment]
From what I see in CellTypes.cs, the following events of the class HyperlinkCell have to be fixed:
- OnMouseEnter: display the hand cursor only if the mouse is over the text of the cell
- OnMouseLeave: display the default cursor as soon as the mouse leave the text
- OnMouseDown: change the textcolor only if the mouse is over the text of the cell
- OnMouseUp: Perform click and change textcolor only if the mouse is over the text of the cell

Of course perform the default behaviour of CellBody otherwise.

I've been playing only for a few hours with the component so I'm not ready yet to propose a real solution. IE: I don't know how to check if the mouse is over the text

**Comment by lamazze** [comment]
@jingwood Hello, I would propose a solution however I have no idea how to evaluate the size of the text inside the cell.
I see in the source several calls like MeasureString, MeasureText or even MeasureCellText which would help comparing the cell bounds to the text bounds in order to check if the mouse click is actually over the text part of the cell.
But I have no idea how to use from inside, for instance, OnMouseUp in HyperlinkCell.
Is it possible as is ?
Thank you!
Fabio

**Comment by jingwood** [comment]
Hi @lamazze, 

I think this requirement can be implemented by measure text, but the problem is how to quickly find a mouse click inside whether or not any range of cell text. It is a problem on the performance side.

For example, the cell A1 has a text that overlays following 10000 cells to A10000. When a mouse click happens on cell A9999, it is necessary to find more than 9999 cells until A1. There must a search algorithm to accelerate the calculation.

ReoGrid uses the viewport concept to manage the current view of cells. Firstly when the viewport updated, like scrolled or zoomed, all overflowed text ranges should be filtered out and marked is whether intersected with the current viewport. When a mouse click happens, check the click is whether inside any marked text ranges.

<img width="678" alt="Screen Shot 2020-02-29 at 10 40 30" src="https://user-images.githubusercontent.com/3468046/75598784-c3f55c80-5ae1-11ea-9173-cd38cca6ad3a.png">

**Comment by lamazze** [comment]
@jingwood I apologize but I don't understand your answer.
What I meant is: if your have a look at the code for OnMouseUp for HyperlinkCell:

```
public override bool OnMouseUp(CellMouseEventArgs e)
{
    if (this.IsPressed)
    {
        if (this.Bounds.Contains(e.RelativePosition))
        {
            this.PerformClick();
        }
    }

    this.IsPressed = false;

    e.Cell.Style.TextColor = VisitedColor;

    return true;
}

```
For me the problem lies with this.Bounds.Contains(e.RelativePosition) which must be replaced by this: <TextBounds>.Contains(e.RelativePosition)

So my question was: how to get the TextBounds from that event ?

I'm not sure I'm not totally wrong with the idea though.

And nice idea btw the "viewport".

Best regards,

Fabio

**Comment by lamazze** [comment]
I've done a proposition. Not sure I've done it the right way.
Let me know what you think about it.
Best regards,
Fabio

**Comment by lamazze** [comment]
@jingwood do you happen to know why, once the left mouse button is down, the component stops receiving mouse move events ? I wanted to mimic the Excel behavior: if the mouse is down, and you move the mouse before button up, the whole process is aborted.
